python/tank/platform/environment_includes.py

```diff
@@ -267,29 +267,84 @@ def find_framework_location(file_name, framework_name, context):
-    
     found_file = None
-    
+    found_token = token
+
     for include_file in include_files:
-                
         # path exists, so try to read it
         included_data = g_yaml_cache.get(include_file) or {}
         
         if token in included_data:
-            found_file = include_file
-        
-    return found_file
-    
+            # If we've been asked to ensure an absolute location, we need
+            # to do some extra work, which might involve recursing up the
+            # config stack until we get to a location descriptor dictionary.
+            if absolute_location:
+                token_data = included_data[token]
+                include_token = None
+
+                # If the value of the token is an include, then we can
+                # recurse up, directly referencing the include name as
+                # the new token.
+                if isinstance(token_data, basestring) and token_data.startswith("@"):
+                    include_token = token_data
+                else:
+                    # In the case where the data isn't itself an include,
+                    # we also need to make sure that the location descriptor
+                    # is itself not an include. If it is, then we still have
+                    # to recurse up the stack until we find the absolute
+                    # location descriptor. In that case, the location value
+                    # becomes the token we're looking for.
+                    if constants.ENVIRONMENT_LOCATION_KEY in included_data[token]:
+                        # Check to see if there's a location descriptor. If there
+                        # is then we need to check to see if that's an include.
+                        location = included_data[token][constants.ENVIRONMENT_LOCATION_KEY]
+                        if location and isinstance(location, basestring) and location.startswith("@"):
+                            include_token = location
+
+                # If we have an include we need to resolve, we take the current
+                # file where we found the include, and the included name becomes
+                # the new token we're looking for.
+                if include_token is not None:
+                    found_file, found_token = find_reference(
+                        include_file,
+                        context,
+                        include_token[1:],
+                        absolute_location=True,
+                    )
+                else:
+                    # We're at the top and we have the concrete location
+                    # descriptor. We can return the include file and token
+                    # where the data was found.
+                    found_file = include_file
+                    found_token = token
+            else:
+                # We've not been asked to resolve the absolute location, so we
+                # don't do any additional work.
+                found_file = include_file
```

Comment:
nitpick: in terms of readability, it sometimes is easier to follow if you switch the if/else around so that you start with the short clause (currently the else clause) and make the long (currently if cluase) the else clause instead.